quickjs-libc.c

```diff
@@ -3473,10 +3653,114 @@ static JSValue js_os_exec(JSContext *ctx, JSValueConst this_val,
         js_free(ctx, envp);
     }
     return ret_val;
+#else
+    if (envp) js_free(ctx, envp);
+    return ret_val;
+#endif
  exception:
     ret_val = JS_EXCEPTION;
     goto done;
 }
+#endif 
+
+#if defined(_WIN32)  
+/* pipe() -> [read_fd, write_fd] or null if error */
+static JSValue js_os_pipe(JSContext *ctx, JSValueConst this_val,
+                          int argc, JSValueConst *argv)
+{
+    int pipe_fds[2], ret;
+    JSValue obj;
+    HANDLE pipe_handles[2];
+    struct _SECURITY_ATTRIBUTES secur;
+    secur.nLength = sizeof(secur);
+    secur.lpSecurityDescriptor = 0;
+    secur.bInheritHandle = true;
+#define WIN32_PIPE_BUFFSIZE 4096
+    ret = CreatePipe(&pipe_handles[0], &pipe_handles[1], &secur, WIN32_PIPE_BUFFSIZE);
+    if (ret == 0) 
+        return JS_NULL;
+    pipe_fds[0] = _open_osfhandle( (intptr_t) pipe_handles[0], 0 );
+    if (pipe_fds[0] == -1 ) {
+        CloseHandle(pipe_handles[0]);
+        return JS_NULL;
+    }
+    pipe_fds[1] = _open_osfhandle( (intptr_t) pipe_handles[1], 0 );
+    if (pipe_fds[1] == -1 ) {
+        CloseHandle(pipe_handles[1]);
+        return JS_NULL;
+    }    
+    obj = JS_NewArray(ctx);
+    if (JS_IsException(obj))
+        return obj;
+    JS_DefinePropertyValueUint32(ctx, obj, 0, JS_NewInt32(ctx, pipe_fds[0]),
+                                 JS_PROP_C_W_E);
+    JS_DefinePropertyValueUint32(ctx, obj, 1, JS_NewInt32(ctx, pipe_fds[1]),
+                                 JS_PROP_C_W_E);
+    return obj;
+}
+
+/* kill(pid) -> (NULL | err) : works like standard c/linux kill except signal is ignored */
+static JSValue js_os_kill(JSContext *ctx, JSValueConst this_val,
+                          int argc, JSValueConst *argv)
+{
+    int pid;
+    HANDLE ph;
+    DWORD flags = PROCESS_TERMINATE;
+    if (JS_ToInt32(ctx, &pid, argv[0]))
+        return JS_EXCEPTION;
+    ph = OpenProcess(flags, false, (DWORD) pid); 
+    if (!ph) 
+        return JS_NewInt32(ctx, GetLastError());
+    if (TerminateProcess(ph, 0)) 
+        return JS_NewInt32(ctx, 0);
+    int err = GetLastError();
+    return JS_NewInt32(ctx,err);
+}
+
+/* watchpid(pid, blocking) -> ret: < 0 = -error | ret: = 0 still waiting | ret: = pid complete
+     hybrid clone of waitpid that will work with Win32 + GNU systems except there's no status  */
+static JSValue js_os_watchpid(JSContext *ctx, JSValueConst this_val,
+                             int argc, JSValueConst *argv)
+{
+    int pid;
+    HANDLE ph;
+    int block = 0, ret; 
+    DWORD options = 0, flags = PROCESS_QUERY_INFORMATION | SYNCHRONIZE;
```

Comment:
This probably should be:
```suggestion
    DWORD options = 0, flags = PROCESS_QUERY_LIMITED_INFORMATION | SYNCHRONIZE;
```

Reply:
tested and works fine like that. I'll put that in. There's better chances of it not tripping security checkpoints in someone else's random section of code with LTD INFO.

Reply:
apparently you close PID HNDS. Unlike FDs to HND casts.

Reply:
I've tested this a good bit over here. I think the handle numbers are recycling, just not immediately. Everything works.